Issue/3726 handle card errors

malinajirka opened the pull request:
Parent issue #3726 

Merge instructions:
1. ~Make sure https://github.com/woocommerce/woocommerce-android/pull/4021 is merged~
2. ~Remove "do not merge" label~
3. Merge this PR

This PR adds support for:
1. Showing "Card declined" error on the UI
2. Showing "No internet connection" error on the UI
3. Accepting "Payment already captured" error from capturePayment endpoint as a success

- Refactors different errors in CardPaymentStatus into a single PaymentFailed with enum indicating its type.
- Introduces PaymentErrorMapper which takes care of converting different types of errors from different sources into CardPaymentStatus.PaymentFailed
- Connects the errors to the UI + fixes buggy layout of the payment dialog
- Adds an extra check to "acceptPayment" and emits an error if StripeTerminal is not initialized

Notes:
- Ignore styles on the UI -> we plan to polish the UI and ask a designer for review when we are done with everything else
- I realize this PR is quite big, but the changes are very related (except of two lines). Also a lot of the changes are newly added unit tests or layout changes which don't need to be reviewed so I hope it won't be too complex to review. However, let me know if you'd prefer me to somehow split it.

To Test:
- Use the simulated reader as it's not possible to simulate payments errors using a hw reader
- Your testing site needs to have WCPay in dev mode installed and activated
1. Open App settings
2. Tap on Manage card reader
3. Check "Use simulated reader" checkbox
4. Tap on "Connect to reader" button
5. Go back and open a detail of an order which is in US dollars
6. Connect debugger (you can still use CI build)
7. Add breakpoint to [this line](https://github.com/woocommerce/woocommerce-android/blob/9738166d32da1b82637f0461d4fea46bc1a0613b/libs/cardreader/src/debug/java/com/woocommerce/android/cardreader/internal/payments/PaymentManager.kt#L61). Make it non-suspandable (uncheck "suspend" checkbox) and check "Evaluate and log" and enter `amountInSmallestCurrencyUnit = 101`. <img width="1063" alt="Screenshot 2021-05-12 at 15 11 16" src="https://user-images.githubusercontent.com/2261188/118038130-59ba8780-b334-11eb-8e80-bd318e10ac6d.png">
8. Tap on "Collect payment" button
9. Notice the payment fails with "Card declined" error
10. Disable the breakpoint and cancel the payment flow (dismiss the dialog)
11. Turn on the airplane mode
12. Tap on "Collect payment"
13. Notice the payment fails with "no network" error
14. Turn off the airplane mode
15. Tap on "Retry"
16. Notice the payment succeeds

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

malinajirka committed:
Refactor PaymentFailed error statuses

malinajirka committed:
Map errors from capture payment endpoint

malinajirka committed:
Move mapping for payment errors to PaymentErrorMapper

malinajirka committed:
User PaymentErrorMapper in PaymentManager

malinajirka committed:
Log StripeTerminalSDK messages only in Debug builds

malinajirka committed:
Add details ProcessPayment log

malinajirka committed:
Fix initialization of CardReaderManager

malinajirka committed:
Bump fluxc to use updated capture payment endpoint

malinajirka committed:
Add support for showing no_network and payment_declined errors

malinajirka committed:
Convert payment layout from constraint layout to linear layout

malinajirka committed:
Remove unused imports

malinajirka commented on WooCommerce/src/debug/kotlin/com.woocommerce.android/WooCommerceDebug.kt:
```diff
@@ -31,8 +38,17 @@ class WooCommerceDebug : WooCommerce() {
-            return !payStore.capturePayment(selectedSite.get(), paymentId, orderId).isError
+        override suspend fun capturePaymentIntent(orderId: Long, paymentId: String): CapturePaymentResponse {
+            val response = payStore.capturePayment(selectedSite.get(), paymentId, orderId)
+            return when (response.error?.type) {
```
We need to map the result from fluxc to CapturePaymentResponse which lives in the CardReaderModule since the module doesn't have FluxC as a dependency -> that's why we use 1:1 mapping.

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt:
```diff
@@ -174,12 +162,17 @@ class CardReaderPaymentViewModel @Inject constructor(
             headerLabel = R.string.card_reader_payment_payment_failed_header,
-            // TODO cardreader use a different label based on the type of the error
-            paymentStateLabel = R.string.card_reader_payment_failed_unexpected_error_state,
+            paymentStateLabel = when (errorType) {
```
I'm not sure if it's a good practice to have `when` in ViewState. But I kind of like when the states are strictly defined so no-one can pass a wrong argument. Wdyt?

kidinov replied:
You can a private extension function here probably, but I personally don't see anything bad in the way you've implemented it

malinajirka commented on WooCommerce/src/main/res/layout/fragment_card_reader_payment.xml:
```diff
@@ -1,11 +1,13 @@
 <?xml version="1.0" encoding="utf-8"?>
-<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
You can ignore these changes in this PR -> we plan to re-work the UI and fix all the styles when we are done with the logic.

malinajirka commented on build.gradle:
```diff
@@ -84,7 +84,7 @@ task installGitHooks(type: Copy) {
 
 ext {
-    fluxCVersion = '9ca3be8e6b5e8a88cd50f0e7430589a0178e552e'
+    fluxCVersion = '96325614198e52afd50c5d812f6703d3ec5274bc'
```
Specific errors for capturePayment endpoints were implemented in https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1992. We didn't merge FluxC back then since the PR didn't introduce any breaking changes.

malinajirka commented on libs/cardreader/src/debug/java/com/woocommerce/android/cardreader/internal/CardReaderManagerImpl.kt:
```diff
@@ -61,9 +62,7 @@ internal class CardReaderManagerImpl(
-            // TODO cardreader: Set LogLevel depending on build flavor.
-            // Choose the level of messages that should be logged to your console
-            val logLevel = LogLevel.VERBOSE
+            val logLevel = if (BuildConfig.DEBUG) LogLevel.VERBOSE else LogLevel.ERROR
```
This is not related to this PR, but I wanted to make 100% sure we will not forget to make this change before the release -> StripeTerminalSDK logs even private data so we need to make sure not to log them in produciton.

malinajirka commented on libs/cardreader/src/debug/java/com/woocommerce/android/cardreader/internal/payments/PaymentManager.kt:
```diff
@@ -147,11 +144,11 @@ internal class PaymentManager(
-        val success = cardReaderStore.capturePaymentIntent(orderId, paymentIntent.id)
-        if (success) {
+        val captureResponse = cardReaderStore.capturePaymentIntent(orderId, paymentIntent.id)
+        if (captureResponse == SUCCESS || captureResponse == PAYMENT_ALREADY_CAPTURED) {
```
If the capturePayment endpoint responds with "PaymentAlreadyCaptured" we can consider it as a success since the user was charged for this payment -> it can for example happen, when the server responds with "SUCCESS" but this response never arrives and the user retries.

malinajirka commented on libs/cardreader/src/test/java/com/woocommerce/android/cardreader/internal/payments/PaymentManagerTest.kt:
```diff
@@ -216,48 +241,15 @@ class PaymentManagerTest {
     }
 
-    @Test
-    fun `given exception contains PaymentIntent,when collecting payment fails, then PaymentData contain that intent`() =
```
These scenarios are test in PaymentErrorMapperTest

malinajirka commented on libs/cardreader/src/test/java/com/woocommerce/android/cardreader/internal/payments/PaymentManagerTest.kt:
```diff
@@ -283,47 +275,15 @@ class PaymentManagerTest {
     }
 
-    @Test
-    fun `given exception contains PaymentIntent,when processing payment fails, then PaymentData contain that intent`() =
```
These scenarios are test in PaymentErrorMapperTest

malinajirka committed:
Update CardReaderPaymentViewModelTest

malinajirka committed:
Merge branch 'issue/3726-migrate-to-hilt' into issue/3726-handle-card-errors

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt

malinajirka committed:
Fix lint

kidinov commented on libs/cardreader/src/debug/java/com/woocommerce/android/cardreader/internal/payments/PaymentErrorMapper.kt:
```diff
@@ -0,0 +1,55 @@
+            TerminalErrorCode.CARD_READ_TIMED_OUT -> CARD_READ_TIMED_OUT
+            TerminalErrorCode.PAYMENT_DECLINED_BY_STRIPE_API -> PAYMENT_DECLINED
+            TerminalErrorCode.REQUEST_TIMED_OUT -> NO_NETWORK
+            else -> CardPaymentStatusErrorType.GENERIC_ERROR
```
`CardPaymentStatusErrorType` is imported, so can be ommited

malinajirka replied:
👍 Fixed in 63bb3ee7ea84ec05824cb0fc9a55212527c45800

kidinov commented on libs/cardreader/src/debug/java/com/woocommerce/android/cardreader/internal/payments/PaymentErrorMapper.kt:
```diff
@@ -0,0 +1,55 @@
+            CapturePaymentResponse.GENERIC_ERROR,
+            CapturePaymentResponse.MISSING_ORDER,
+            CapturePaymentResponse.CAPTURE_ERROR,
+            CapturePaymentResponse.SERVER_ERROR -> CardPaymentStatusErrorType.GENERIC_ERROR
```
and here as well

malinajirka replied:
👍 Fixed in 63bb3ee7ea84ec05824cb0fc9a55212527c45800

kidinov commented on libs/cardreader/src/debug/java/com/woocommerce/android/cardreader/internal/payments/PaymentErrorMapper.kt:
```diff
@@ -0,0 +1,55 @@
+    ): PaymentFailed {
+        val paymentData = PaymentDataImpl(originalPaymentIntent)
+        val message = "Capturing payment failed: $capturePaymentResponse"
+        val type = when (capturePaymentResponse) {
```
I think we could benefit here from the use of [exhaustive](https://proandroiddev.com/til-when-is-when-exhaustive-31d69f630a8b) approach.

malinajirka replied:
Interesting solution to this issue - I've encountered it several times, but it never came to my mind to use a code as a solution -> I always tried to setup lint/checkstyle/... to cover this scenario. It's definitely worth considering. However, I believe it doesn't apply to this particular case as it is already an expression. Wdyt?

kidinov replied:
Yeah, my bad. I was probably referring to another place =)

kidinov commented on libs/cardreader/src/debug/java/com/woocommerce/android/cardreader/internal/payments/PaymentManager.kt:
```diff
@@ -27,32 +24,38 @@ import com.woocommerce.android.cardreader.internal.payments.actions.CreatePaymen
             return@flow
         }
-        val amountInSmallestCurrencyUnit = try {
+        var amountInSmallestCurrencyUnit = try {
```
Shouldn't it be `val` here?

malinajirka replied:
Good catch, I changed it to "var" for testing purposes -> I wanted to be able to use the debugger to change the value and I forgot to revert it.

malinajirka replied:
Fixed in b33c5307e0d0806c91c59d97616c3b83a896e912

kidinov commented on libs/cardreader/src/debug/java/com/woocommerce/android/cardreader/internal/payments/PaymentManager.kt:
```diff
@@ -27,32 +24,38 @@ import com.woocommerce.android.cardreader.internal.payments.actions.CreatePaymen
+            emit(errorMapper.mapError(errorMessage = "Reader not connected"))
             return@flow
         }
         var paymentIntent = createPaymentIntent(amountInSmallestCurrencyUnit, currency)
```
Also, `val` probably?

malinajirka replied:
👍 Fixed in b33c5307e0d0806c91c59d97616c3b83a896e912

kidinov commented on libs/cardreader/src/debug/java/com/woocommerce/android/cardreader/internal/payments/PaymentErrorMapper.kt:
```diff
@@ -0,0 +1,55 @@
+            CapturePaymentResponse.CAPTURE_ERROR,
+            CapturePaymentResponse.SERVER_ERROR -> CardPaymentStatusErrorType.GENERIC_ERROR
+            CapturePaymentResponse.PAYMENT_ALREADY_CAPTURED,
+            CapturePaymentResponse.SUCCESS ->
```
Having a type hierarchy something like that would move this check to the compile-time.

Also, in places like [that](https://github.com/woocommerce/woocommerce-android/pull/4022/files#diff-9cca462298f606e68a2a3c5ac516b7d95d1c89cebdbf6f7a0efa159c7739f212R148) you don't need to modify code  letter if new error or success will be handled

```
    sealed class CapturePaymentResponse {
        sealed class Successful : CapturePaymentResponse() {
            object SUCCESS : Successful()
            object PAYMENT_ALREADY_CAPTURED : Successful()
        }

        sealed class Error : CapturePaymentResponse() {
            object GENERIC_ERROR : Error()
            object PAYMENT_ALREADY_CAPTURED : Error()
            object MISSING_ORDER : Error()
            object CAPTURE_ERROR : Error()
            object SERVER_ERROR : Error()
            object NETWORK_ERROR : Error()
        }
    }
```

malinajirka replied:
❤️ I completely agree that the proposed structure is nicer and I like it a lot. I'd like to share my thoughts with you as I'm not sure how to make such change without breaking responsibilities.

I believe that FluxC shouldn't decide if `PAYMENT_ALREADY_CAPTURED` is a success or not, since it's a business logic. So the mapping should take place somewhere in the apps repo. One possible candidate would be [this place](https://github.com/woocommerce/woocommerce-android/blob/ba1ba33362bc405699cede8159c1f290d7d0c57b/WooCommerce/src/debug/kotlin/com.woocommerce.android/WooCommerceDebug.kt#L43). However, it currently contains mapping  that exists just so we can pass FluxC specific model to CardReaderModule which doesn't have FluxC as a dependency -> I believe that business logic doesn't belong here either. This leaves us with keeping the mapping somewhere in CardReaderModule. We could add extra line [here](https://github.com/woocommerce/woocommerce-android/blob/b33c5307e0d0806c91c59d97616c3b83a896e912/libs/cardreader/src/debug/java/com/woocommerce/android/cardreader/internal/payments/PaymentManager.kt#L147) and map `CapturePaymentResponse` to a newly introduce model, but this mapping would still need to have lines such as `captureResponse == SUCCESS || captureResponse == PAYMENT_ALREADY_CAPTURED` and introducing yet another model feels like an overkill.


What do you think about all this? Do you think it'd be ok to keep this logic in FluxC or WooCommerceDebug (we could extract it into a standalone mapper of course)? Or do you think my last point isn't an overkill and it's worth introducing a new model + new mapping?

kidinov replied:
I would go with just a standalone mapper which is used in `WooCommerceDebug`.  Maybe it's a bit of business logic is leaking there, but the model itself defines this business logic so probably not a big deal anyway

kidinov replied:
Btw, https://kotlinlang.org/docs/whatsnew15.html#sealed-interfaces may be handy here. For some reason, we are not on kotlin 1.5 yet, but I think it's should be easy to update

kidinov reviewed (commented):
Looks good, just couple of things which we might consider to improve

kidinov commented:
What I noticed is that I test it without internet connection I get "GENERIC_ERROR", not the network error

malinajirka committed:
Remove unnecessary explicit path

malinajirka committed:
Minor refactorings

malinajirka committed:
Use explicit method names in PaymentErrorMapper

malinajirka committed:
Use mapTerminalError when creating payment intent fails

malinajirka committed:
Merge branch 'develop' into issue/3726-handle-card-errors

# Conflicts:
#	WooCommerce/src/debug/kotlin/com.woocommerce.android/WooCommerceDebug.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt
#	WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModelTest.kt

malinajirka committed:
After merge fix

malinajirka commented:
Thanks so much for the review @kidinov! 

> What I noticed is that I test it without internet connection I get "GENERIC_ERROR", not the network error

Ahh, you are right -> when I tested this flow I always turned the airplane mode on, after the collect payment step. Fixed in 350fa1ecb3d644a447feebd0cd0bd8bae4d68514 and be627a583a7917093fb8ed8d7de8d8cb8fbcaa33.

malinajirka committed:
Refactor CapturePaymentResponse to improve readability and type safety

malinajirka committed:
Map FluxC response to CardReaderModule response

malinajirka committed:
Use updated CapturePaymentResponse in PaymentManager and PaymentErrorMapper

malinajirka committed:
Use Root locale to convert currency to lower case

malinajirka committed:
Merge branch 'develop' into issue/3726-handle-card-errors

# Conflicts:
#	build.gradle

malinajirka commented:
@kidinov I've updated CapturePaymentResponse - great suggestion, the code is definitely cleaner and less error prone now. I've also fixed a small issue in https://github.com/woocommerce/woocommerce-android/pull/4022/commits/60d1eac40318a50c190c9f84d14f7d48645e6270 and after seeing your solution I changed it to https://github.com/woocommerce/woocommerce-android/pull/4022/commits/95ed4040c60a37f46d00c2f1766999d9219c5c51. It's ready for another round. Thanks!

malinajirka committed:
Replace toLowerCase with equals(..., ignoreCase = true)

malinajirka committed:
Remove unused import

kidinov commented on libs/cardreader/src/debug/java/com/woocommerce/android/cardreader/internal/payments/PaymentManager.kt:
```diff
@@ -144,11 +144,9 @@ internal class PaymentManager(
-        } else {
-            emit(errorMapper.mapCapturePaymentError(paymentIntent, captureResponse))
+        when (val captureResponse = cardReaderStore.capturePaymentIntent(orderId, paymentIntent.id)) {
+            is CapturePaymentResponse.Successful -> emit(PaymentCompleted)
```
👍

kidinov reviewed (approved):
👍
